Feature/line mvp

Neal-Y opened the pull request.

Neal-Y committed:
migration change

Neal-Y committed:
frontend index

Neal-Y committed:
go mod

Neal-Y committed:
mysql utf8 fixed

Neal-Y committed:
post format

Neal-Y committed:
add handler for test

Neal-Y committed:
user model crud

Neal-Y committed:
user controller auth

Neal-Y committed:
user router

Neal-Y committed:
feature (line mvp use case) go mod & go sum

Neal-Y committed:
feature (line mvp use case) go mod & go sum

Neal-Y committed:
feature (line mvp use case) main.go load config

Neal-Y committed:
feature (line mvp use case) config setting

Neal-Y committed:
feature (line mvp constant) create constant

Neal-Y committed:
feature (line mvp constant) refactor auth process and hide sensitive information

Neal-Y committed:
test

Neal-Y committed:
fix (.gitignor) .env & .gitignore

Neal-Y committed:
Ignore vendor directory

Neal-Y committed:
Delete .DS_Store

Neal-Y committed:
feat(entity)更改db logic移到repository layer

Neal-Y committed:
feat(repository)將原本在entity中的crud邏輯，做些更改並且放在repository layer

jonyig commented on handler/general/index.go:
```diff
@@ -27,7 +24,11 @@ func newRoute(h *General, r *gin.RouterGroup) {
+}
 
+func (h *General) test(c *gin.Context) {
+	c.HTML(http.StatusOK, "index.html", nil)
```
test 的 api 是測試的嗎

Neal-Y replied:
對，很久以前要準備開始寫line login複習一下用的😂

Neal-Y committed:
feat(entity)添加 user 的 dto

Neal-Y committed:
feat(repository)添加所需的db logic

Neal-Y committed:
feat(service)將db logic組合起來供controller使用，並且預留簡單的CRUD函數

Neal-Y committed:
feat(controller)將冗余的business logic移到service layer，同時保持controller的簡潔

Neal-Y committed:
feat(controller)添加controller 中的service & repository的初始化

jonyig commented on handler/user/auth.go:
```diff
@@ -0,0 +1,45 @@
+	state := c.Query("state")
+
+	if state != "randomStateString" {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid state"})
```
這邊有任何的處理失敗，都需要導到一個頁面吧，好讓使用者可以使用其他登入方式吧

Neal-Y replied:
我想說你這個地方說是impv，就是目前先假設line不會爆炸，不然就是直接mac id或是其他可以驗證使用者的方式，不過我是沒什麼頭緒要怎麼弄ＱＱ

jonyig replied:
可以先導到一個頁面吧，聯絡管理員之類的

Neal-Y replied:
[dcdb719](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/dcdb7199a57dfd1b0e9ceb25bef9ce8d283d18f9)

jonyig commented on service/user.go:
```diff
@@ -0,0 +1,129 @@
+		return "", err
+	}
+
+	var tokenData map[string]interface{}
```
定義一個 struct 出來用吧

Neal-Y replied:
[21081b3](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/21081b3273394827c7edcafbf282e8045771ae2c)

jonyig commented on service/user.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (s *userService) GetLineProfile(accessToken string) (*database.User, error) {
+	req, err := http.NewRequest("GET", constant.LineProfileURL, nil)
```
這邊跟上面的 http call out 操作，可以整理一起嗎？

Neal-Y replied:
你指的是兩個函數合併成一個嗎？

Neal-Y replied:
[f295995](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/f295995abe5c1537ee8153bdb3319e5d14dbb7f6)

Neal-Y committed:
刪除無用post相關檔案

Neal-Y committed:
feat(controller)刪除post在route的config

Neal-Y committed:
feat(service)將原本兩個token以及Profile的函數合併成一個

Neal-Y committed:
feat(entity)將service中的struct整理成一個dto提取到entity layer做紀錄

Neal-Y committed:
merge

Neal-Y committed:
feat(util)把service中重複的http call out提取出來

Neal-Y committed:
feat(frontend)error.html

Neal-Y committed:
feat(util)新增parse IDtoken的函數，將userid、displayname、email從IDtoken中解jwt提取出來

Neal-Y committed:
feat(constant)刪除無用lineprofileURL

Neal-Y committed:
go mod go sum

Neal-Y committed:
feat(builder)使用builder pattern將原本在service中的冗余過程提取出來

Neal-Y committed:
feat(service)refactor並且將重複的json解析和token parse移出去到util

Neal-Y committed:
feat(handler/user)將網頁綁定到api上

Neal-Y committed:
feat(handler/user)如果失敗將由handleLineServerError()處理打印出log錯誤，並且redirect到error.html

Neal-Y committed:
feat(route)將loadhtml移到handler/user中的index.go

Neal-Y committed:
fix(route)重複的loadHTML函數刪除

Neal-Y committed:
feat(util)移除掉Get()和PostForm()兩個函數，將他們整合在httpBuilder當中

Neal-Y committed:
feat(builder)由於參數太過客製，我們新增了builder pattern 替代

Neal-Y committed:
feat(service)使用builder pattern替代

jonyig commented on builder/http.go:
```diff
@@ -0,0 +1,60 @@
+	}
+}
+
+func (b *HTTPClient) WithMethod(method string) *HTTPClient {
```
nitpick:

這邊我習慣限縮我想要開放的 method

like 
```
withMethodGet()
withMethodPost()
withMethodPut()
```

Neal-Y replied:
你說避免method送錯導致不必要的api報錯？

Neal-Y replied:
[a49e649](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/a49e649e22f1aafcac39013ddf667d9978ed878b)

Neal-Y replied:
fix(builder)限制給出的method

jonyig commented on service/user.go:
```diff
@@ -0,0 +1,102 @@
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
```
這邊兩個是為何？

Neal-Y replied:
沒清乾淨ＸＤＤ

Neal-Y replied:
[a0b0fda](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/a0b0fda4921305d779d2af480e67bac894f1092e)

Neal-Y replied:
[be225c2](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/be225c27e3b2a9a2abe2beb3edaa3c23e64004f7)

jonyig commented on service/user.go:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	var tokenData *datatransfer.LineTokenResponse
+	err = util.ParseJSONResponse(body, &tokenData)
```
這段也可以嘗試塞進去 http client builder 裏面

Neal-Y replied:
[be225c2](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/be225c27e3b2a9a2abe2beb3edaa3c23e64004f7)

jonyig commented on builder/http.go:
```diff
@@ -0,0 +1,60 @@
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
```
針對 header 處理我會開出幾種彈性
當前這邊我會這樣用
```
UserHeaderFormUrlencoded()
```

可以預期未來這個很容易出現，先開出來不會有什麼超大成本
```
SetHeader(map)
```

Neal-Y replied:
[ddbe16f](https://github.com/Neal-Y/devmentor-BE103-golang/pull/5/commits/ddbe16fe2a2d25ba711a0f7fa53ae11eea25b679)

Neal-Y committed:
product 刪除util移出來與order共用，同時將所有用到這函數指定從哪個pkg出來的

Neal-Y committed:
fix(builder)限制給出的method

Neal-Y committed:
fix(service)針對重複的if err != nil做修改，並且更新新的builder處理method的函數到ExchangeTokenAndGetProfile中

jonyig commented on builder/http.go:
```diff
@@ -0,0 +1,65 @@
+package builder
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
```
Test

Neal-Y committed:
fix(builder)更改以下幾點：
1.新增擴充函數SetHeader()、以及當前要用的UserHeaderFormUrlencoded()
2.新增泛型(generic)來取代Build在吃interface參數時的各種副作用
3.將Parse Json這件事情從service layer轉嫁到builder中處理，保持service layer的乾淨

Neal-Y committed:
fix(builder)修改ExchangeTokenAndGetProfile()的內容將它丟到builder中，以及新增泛型

Neal-Y committed:
fix(builder)忘記初始化header的map
